File: packages/transit-py-runtime/transit_service.py
"""
Example Transit Python service.

This file demonstrates how to create a Python service that can be called from JS.
Copy this file to your project and modify the functions as needed.
"""

import json
from transit_server import TransitServer, register_function


def process_data(args_json):
    """Process data and return a result.

    Args:
        args_json: JSON string with arguments from JS

    Returns:
        JSON string with the result
    """
    args = json.loads(args_json)
    data = args.get("data", [])

    return json.dumps(
        {
            "output": f"Python processed {len(data)} items",
            "processed": True,
            "items": len(data),
        }
    )


def get_version(args_json):
    """Return the service version.

    Args:
        args_json: JSON string (ignored)

    Returns:
        JSON string with version info
    """
    return json.dumps({"version": "1.0.0", "runtime": "python"})


def transform_text(args_json):
    """Transform text data.

    Args:
        args_json: JSON string with "text" and optional "operation" fields

    Returns:
        JSON string with transformed text
    """
    args = json.loads(args_json)
    text = args.get("text", "")
    operation = args.get("operation", "upper")

    if operation == "upper":
        result = text.upper()
    elif operation == "lower":
        result = text.lower()
    elif operation == "reverse":
        result = text[::-1]
    else:
        result = text

    return json.dumps({"result": result, "operation": operation})


# ─── Main entry point ─────────────────────────────────────────────────────────

if __name__ == "__main__":
    server = TransitServer()

    # Register functions
    register_function("processData", process_data)
    register_function("getVersion", get_version)
    register_function("transformText", transform_text)

    # Start the server (blocks until stopped)
    server.start()

File: demo/java/src/main/java/lib/transit/java/TransitService.java
package transit.java;

/**
 * Sample Transit Java service.
 * User code implements functions that accept/return JSON strings.
 */
public class TransitService {

    /**
     * Example: process a job and return a result.
     * Called from JS via transit.
     */
    public String processSpecialized(String argsJson) {
        // Simple manual JSON construction for zero dependencies
        // In real code, users bring their own JSON library
        return "{\"id\":" + extractField(argsJson, "id") +
               ",\"output\":\"Java processed " + countArrayElements(argsJson) + " bytes\"" +
               ",\"processed\":true}";
    }

    public String getVersion(String argsJson) {
        return "{\"version\":\"0.1.0-java\"}";
    }

    // ─── Minimal JSON helpers (no external deps) ──────────────────────────

    private static String extractField(String json, String field) {
        String key = "\"" + field + "\"";
        int idx = json.indexOf(key);
        if (idx < 0) return "null";
        int colon = json.indexOf(':', idx + key.length());
        if (colon < 0) return "null";
        // Find value start
        int valStart = colon + 1;
        while (valStart < json.length() && json.charAt(valStart) == ' ') valStart++;
        if (json.charAt(valStart) == '"') {
            int valEnd = json.indexOf('"', valStart + 1);
            return json.substring(valStart, valEnd + 1);
        }
        // Number or other literal
        int valEnd = valStart;
        while (valEnd < json.length() && ",}] \n\r\t".indexOf(json.charAt(valEnd)) < 0) valEnd++;
        return json.substring(valStart, valEnd);
    }

    private static int countArrayElements(String json) {
        int idx = json.indexOf("\"bytes\"");
        if (idx < 0) return 0;
        int bracket = json.indexOf('[', idx);
        if (bracket < 0) return 0;
        int close = json.indexOf(']', bracket);
        if (close < 0) return 0;
        String arr = json.substring(bracket + 1, close).trim();
        if (arr.isEmpty()) return 0;
        return arr.split(",").length;
    }

    /**
     * Main entry point — starts the Transit server.
     * Server binds to an ephemeral port on 127.0.0.1 and prints PORT=<port> to stdout.
     */
    public static void main(String[] args) throws Exception {
        TransitServer server = new TransitServer();
        TransitService service = new TransitService();

        // Register functions
        server.registerFunction("processSpecialized", service::processSpecialized);
        server.registerFunction("getVersion", service::getVersion);

        // Start server (blocks until shutdown)
        server.start();
    }
}
